Optional syslog support

troglobit opened the pull request:
This pull request adds an optional syslog backend alternative to the default stdout/stderr logging.  It is primarily intended for systems using RAUC without systemd, e.g., an embedded system built around BusyBox init.

It is of course possible to use RAUC as-is on non-systemd systems, however with this patch users avoid ANSI escape sequences (color) in their logs and can perform severity based filtering for remote logging of critical events.

We use an earlier version of this patch in the [Infix NOS](https://github.com/kernelkit/infix/).  This updated version adds configurable facility support and also ensures to replace the default logger.

troglobit committed:
src/main: add optional syslog support

This patch adds an optional syslog backend alternative to the default
stdout/stderr logging.

The syslog backend is enabled with `--syslog` on the command line.  It
logs directly to the local syslog daemon, via the POSIX syslog() API,
using a default 'daemon' facility.  This facility can be changed if an
optional argument is given, e.g., `--syslog=local0`.  Useful for basic
message filtering to a dedicated log file.

Signed-off-by: Joachim Wiberg <troglobit@gmail.com>

jacmet commented:
LGTM, it would indeed be handy to have rather than having to pipe rauc stdout/sterr to logger

jluebbe commented on src/main.c:
```diff
@@ -10,6 +10,8 @@
 #include <locale.h>
 #include <stdio.h>
 #include <string.h>
+#define SYSLOG_NAMES
```
Is this really needed? https://sourceware.org/bugzilla/show_bug.cgi?id=16355 sounds like we should avoid this define.

troglobit replied:
The define opens up the use of `facilitynames[]`.  It has been a long standing practice, but we can replace that with a local array of the same names if you want.

jluebbe reviewed (commented):
Since 2.80, glib [has native support for syslog](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3791). Currently, we only require glib 2.56. Furthermore, the glib implementation defaults to the "user" facility (if not overridden on the message level). So we can't easily use the native support.

Still, we could take some inspiration from the implementation, such as avoiding the need for `SYSLOG_NAMES` by mapping to the integer levels directly.

Which use-case do you see for specifying the facility on the command line? Perhaps hard-coding it to the `LOG_DAEMON` value would be enough?

troglobit commented:
@jluebbe I take it you are still overall in favor of adding native `syslog()` support in this form?  Or would you prefer closing this in favor of the native (later) glib support?  I'm fine either way, we can carry the patch locally until we get the bump in glib version.

jluebbe commented on src/main.c:
```diff
@@ -2646,6 +2648,67 @@ static void create_option_groups(void)
+		}
+	}
+
+	openlog(G_LOG_DOMAIN, LOG_PID | LOG_NOWAIT, facility);
```
What's the effect of `LOG_NOWAIT` with sysklogd?

troglobit replied:
Internally it will connect immediately to the `/dev/log` socket instead of postponing it until the first `syslogd()` call.

jluebbe commented:
> @jluebbe I take it you are still overall in favor of adding native `syslog()` support in this form? Or would you prefer closing this in favor of the native (later) glib support? I'm fine either way, we can carry the patch locally until we get the bump in glib version.

To full replace this native implementation with the glib feature, we'd need to add a way to set the default facility via the glib API, which would result in an even higher minimum version. So I'm still in favor of a native implementation.

It would be good to have test coverage for the new code, so that we can avoid regressions. Can we temporarily run a syslog service to catch messages from rauc running with `--syslog`?

jacmet commented:
> To fully replace this native implementation with the glib feature, we'd need to add a way to set the default facility via the glib API, which would result in an even higher minimum version. So I'm still in favor of a native implementation.

glib 2.80 is also quite recent (~10 months ago), so it would be annoying to require it

troglobit commented:
> > @jluebbe I take it you are still overall in favor of adding native `syslog()` support in this form? Or would you prefer closing this in favor of the native (later) glib support? I'm fine either way, we can carry the patch locally until we get the bump in glib version.
> To fully replace this native implementation with the glib feature, we'd need to add a way to set the default facility via the glib API, which would result in an even higher minimum version. So I'm still in favor of a native implementation.

OK, cool thanks!

> It would be good to have test coverage for the new code, so that we can avoid regressions. Can we temporarily run a syslog service to catch messages from rauc running with `--syslog`?

Should definitely be possible.  I'll have a look at your regression test system and see what I can do.

jluebbe commented:
@troglobit Do you need some help for the testing side? If that's blocking you, we could try to find some time to prepare something there.

troglobit commented:
> @troglobit Do you need some help for the testing side? If that's blocking you, we could try to find some time to prepare something there.

Sorry for dragging my feet on this one, I've been reassigned to other projects for a while.  Been meaning to get back to this, and possibly in a week or two, now that Sweden is shutting down for the summer, I'll be back on it.  If I get stuck, I'll let you know, thanks!

troglobit committed:
test/test_syslog: verify optional syslog support

This new test verifies syslog support by creating a dummy listener for
the /dev/log socket and checking the log format sent to this socket to
ensure it contains rauc[PID]: style logs like 'service start'.

Three different variations, for maximum coverage, are run:

 --syslog           (no argument, default facility)
 --syslog=local0    (custom facility, should pass OK)
 --syslog=undefined (invalid facility, should FAIL)

Note, test is skipped if /dev/log already exists or if it cannot bind
the domain socket to /dev/log.  Either a syslog daemon is running, a
case the test is not designed for, or the test is running in a cross
environment without podman being started with --privileged.

Signed-off-by: Joachim Wiberg <troglobit@gmail.com>

troglobit commented:
There, finally some tests for the syslog backend!  I hope they meet your expectations.

Note, the test relies on the standard libc `syslog()` API, which writes to `/dev/log`.  So the test sets up a UNIX socket and tries to bind it to `/dev/log`, if this fails, as it does on the cross tests because of non-privileged mode of `podman`, then the test is skipped.  Otherwise the test emulates a syslog daemon and starts the `rauc service` with different options to verify the intended behavior.

troglobit commented:
Ping :bell: I hope this meets your expectations, @jluebbe :smiley:

edmundoferreira commented:
@troglobit really nice feature, this would be really helpful.

I do have one question though: 

How will this `rauc -s` option play with rauc `system.conf` [`[log.<logger>] section`](https://rauc.readthedocs.io/en/latest/reference.html#log-logger-sections)?
- It overrides the `system.conf` `filename`? 
- Can I specify this in my `system.conf` instead, to have syslog integration insteading of passing `-s` all the time? if yes how?

jluebbe commented:
#1865 also touches a related area (it wants to set the syslog identifier for journald output).

jluebbe commented:
@troglobit What is the (or your) use case for setting the facility name? Is that something you'd customize often?

Why did you add a commandline option for the facility instead of having it in the `system.conf`? Is it because we write messages before loading the config?

troglobit commented:
> @troglobit What is the (or your) use case for setting the facility name? Is that something you'd customize often?

Sorting. On more basic systems, that only have BusyBox syslogd, facilities are the key used for directing output from different subsystems to given log files.  In Infix OS we have this [mapping of localN](https://github.com/kernelkit/infix/blob/main/board/common/rootfs/etc/syslog.d/readme.txt), where rauc gets its own `/var/log/upgrade.log` that's automatically rotated by sysklogd.

> Why did you add a commandline option for the facility instead of having it in the `system.conf`? Is it because we write messages before loading the config?

The original patch for this was made a few years ago, so I cannot say for certain, but it's very plausible I wanted to catch everything regardless of the config file.  However, it's also a practice of mine for other daemons I maintain, so it's equally likely I didn't even stop to consider the config file :sweat_smile:
